### PR TITLE
code-review: update requirements

### PR DESCRIPTION
**Pull request by OleksiyRudenko**
## Major changes:

* add intro to General Requirements

* update Code Formatting section

* add note on variable/function naming (verbs, nouns etc)

* add explanation re Event Delegation

* add requirements on TJSW pre-OOP

* update requirements on OO JS (Frogger)

* update requirements on OOP Exercise

## Next steps

Once this PR finalized and merged:

1. Publish updated docs to kottans.org/documentation.
2. Update [homeworks 2022 checklists](https://github.com/kottans/frontend-2022-homeworks/tree/main/.github/pr-checklists) so that bot's messages on TJSW, Frogger and OOP Exercise lead to kottans.org/documentation sections rather than contain checklists.
3. Announce changes in Students' chat.

### CODE REVIEW

File: docs/doc/code-review.md [hunk 1]
```diff
@@ -3,15 +3,22 @@ id: code-review
...
 - [How to Do Code Reviews Like a Human (Part One)](https://mtlynch.io/human-code-reviews-1/) /
   [Code review по-человечески (часть 1)](https://habr.com/en/post/340550/)
 - [How to Do Code Reviews Like a Human (Part Two)](https://mtlynch.io/human-code-reviews-2/) /
   [Code review по-человечески (часть 2)](https://habr.com/ru/post/342244/)
 - [PR Code Review Phrase Book](https://gist.github.com/OleksiyRudenko/e6f573d7aca2cc854ccce6087cfe7138)
 
-## Formatting, Style, Manual Testing
+## General requirements
+
+The code is written for other people to read and contribute to comfortably.
```
Comment (artem-trubin) [inline docs/doc/code-review.md:16]: `and to contribute comfortably`?

File: docs/doc/code-review.md [hunk 2]
```diff
@@ -25,11 +32,10 @@ PRs should have proper name (as per task name), should contain link to working d
...
-- Please advise your peer to fix irregular indentations and remove redundant empty lines.
-- Please advise your peer to put newline at the end of every file.
-- Unnecessary comments should be avoided. Advise your peer to find a way to express their intent through expressive variable names or by abstracting part of the code into properly named function.
-- Magic numbers are [bad](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)
-- Add an empty lines at the end of each file. [Reason](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline?noredirect=1&lq=1)
+- Fix irregular indentations and remove redundant empty lines.
+- Put an empty line at the end of every file. [Reason](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline?noredirect=1&lq=1). Tune your code editor's settings so it does this for you.
+- Unnecessary comments should be avoided. Find a way to express the intent through expressive variable names or by abstracting part of the code into properly named function. Comments, if there are any, should explain "why", not "what".
```
Comment (artem-trubin) [inline docs/doc/code-review.md:37]: `express` through `expressive`. Maybe `express the intent through descriptive variable names` is better?

Comment (artem-trubin) [reply docs/doc/code-review.md:37]: `into a properly named function` or `into properly named functions`

File: docs/doc/code-review.md [hunk 3]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
 2. This is an imperative way, you need to write declarative and describe what your code does, not how. This will make your code shorter and easier to maintain.
 3. Reuse of code. Saying, you will need to rotate some other stuff - you will add a similar line to another part of an application. Which is not right because of DRY.
 
-- Separation of responsibility - JS for logic, CSS - for styling.
+So, replace such parts with css classes. You can use `classList` to manipulate them.
```
Comment (artem-trubin) [inline docs/doc/code-review.md:283]: `CSS`, uppercase

File: docs/doc/code-review.md [hunk 4]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
 
 `keyUp` handles not all input types (try pasting text via context menu instead of typing)
 
+### A Tiny JS World -- (pre-OOP)
+
+_Relates to [Building a Tiny JS World](https://github.com/kottans/frontend/blob/master/tasks/js-pre-oop.md) task._
+
+1. Men and women belong to the same biological species.
+1. `prototype`-based or ES2015/ES6 `class` syntax aren't used.
+1. Code is [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), which means that whenever you see a pattern in your code those should be eliminated as much as possible. Examples:
```
Comment (artem-trubin) [inline docs/doc/code-review.md:297]: `a pattern` - `it should be eliminated`

File: docs/doc/code-review.md [hunk 5]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
+   1. `print(dog); print(cat); etc ... ` should be refactored employing `Array.forEach` as the least
```
Comment (artem-trubin) [inline docs/doc/code-review.md:298]: No dot in the end.

File: docs/doc/code-review.md [hunk 6]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
-        doesn't **contain** any inline code to check e.g. collisions)
-  - [ ] the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
-- Most common mistakes
-  - [ ] Make sure `target = condition ? valueWhenConditionTrue : valueWhenConditionFalse` is used instead of
-        `condition ? target = valueWhenConditionTrue : target = valueWhenConditionFalse`;
-        [Conditional (ternary) operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Conditional_Operator)
+A. Minimal requirements to meet:
+1. employ ES6 features like `const`, `let` etc. (with exclusion of ES6 `class` syntax)
+1. the code is very [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
+1. Requirements re **Constants**:
```
Comment (artem-trubin) [inline docs/doc/code-review.md:311]: `are`?

File: docs/doc/code-review.md [hunk 7]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
+   1. properties common for some classes are generalized into a base class
+      (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
+   1. class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
+      not `Subclass.prototype = new Superclass(params);`;
+      [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
+   1. classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
+      (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
+      Hint: pass instance of a game object (or objects) as an argument to other game objects they need to interact with
+   1. Separation of Concerns principle is followed
+      (e.g. `update` method does only rendering and
```
Comment (artem-trubin) [inline docs/doc/code-review.md:331]: But `update` method shouldn't render anything. It is for update the state of the object which is then used for rendering in another method. Maybe a bad example?

File: docs/doc/code-review.md [hunk 8]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
 
-Minimal requirements to meet:
-
-- [ ] Implement a base class to inherit from
-- [ ] Employ default parameters
-- [ ] Each species is represented with its own class
-- [ ] No need to specify species at instantiation
-- [ ] Classes for species that do not have hands by natural design
+1. Minimal requirements to meet:
+   1. Implement base classes the child classes inherit shared properties from
```
Comment (artem-trubin) [inline docs/doc/code-review.md:344]: Better wording:
`Implement base classes for child classes to inherit shared properties from`.

File: docs/doc/code-review.md [hunk 9]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
+   1. all numbers like block dimensions, initial locations are defined as constants
+   1. every number that has a semantic purpose (like those listed above) should be defined as constants; think of how your code **reads** - the closer to plain English the better
+   1. there are core constants and derived (calculated) constants
+      (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
+   1. arrays of constants also qualify as constants
+      (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
+   1. const objects help organizing and structure const data even better
+      (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
+1. Requirements re **OOP**:
+   1. OO is implemented using JS prototype chain object model (**not** ES2015/ES6 `class` syntax)
```
Comment (artem-trubin) [inline docs/doc/code-review.md:321]: We have to figure out do we really need to restrict `class` usage for students, when some mentors are for it.

File: docs/doc/code-review.md [hunk 10]
```diff
@@ -3,15 +3,22 @@ id: code-review
 title: How to Conduct A Helpful Code Review
 ---
 
-Consider following links:
+As a reviewer consider the following guidelines:
 
 - [How to Do Code Reviews Like a Human (Part One)](https://mtlynch.io/human-code-reviews-1/) /
   [Code review по-человечески (часть 1)](https://habr.com/en/post/340550/)
```
Comment (artem-trubin) [inline docs/doc/code-review.md:9]: Maybe remove `habr` links? Russian site, doesn't work without VPN.

Comment (OleksiyRudenko) [reply docs/doc/code-review.md:9]: Makes sense as we have links to original posts anyway and code reviewers are likely have sufficient command of English.
Cannot remember who suggested adding translations and why. Would expect yet another opinion on this matter.

File: docs/doc/code-review.md (hunk 1 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:16]: Suggesting to add a preposition and an excplicit reference to the object of contribution, for the sake of clarity.
```suggestion
The code is written for other people to read and to contribute to the code comfortably.
```

File: docs/doc/code-review.md (hunk 2 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:37]:
```suggestion
- Unnecessary comments should be avoided. Find a way to express the intent through descriptive variable names or by abstracting part of the code into properly named function. Comments, if there are any, should explain "why", not "what".
```

File: docs/doc/code-review.md (hunk 3 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:283]:
```suggestion
So, replace such parts with CSS classes. You can use `classList` to manipulate them.
```

File: docs/doc/code-review.md (hunk 4 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:297]:
```suggestion
1. Code is [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself), which means that whenever you see a pattern in your code it should be eliminated. Examples:
```

File: docs/doc/code-review.md (hunk 5 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:298]:
```suggestion
   1. `print(dog); print(cat); etc ... ` should be refactored employing `Array.forEach` as the least.
```

File: docs/doc/code-review.md [hunk 11]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
+1. Requirements re **Constants**:
+   1. all numbers like block dimensions, initial locations are defined as constants
+   1. every number that has a semantic purpose (like those listed above) should be defined as constants; think of how your code **reads** - the closer to plain English the better
+   1. there are core constants and derived (calculated) constants
+      (e.g. `const FIELD_WIDTH = BLOCK_WIDTH * BLOCKS_NUMBER;`)
+   1. arrays of constants also qualify as constants
+      (e.g. `const INITIAL_POSITIONS = [1,2,3,4].map(rowNumber => rowNumber * BLOCK_HEIGHT);`)
+   1. const objects help organizing and structure const data even better
+      (e.g. `const PLAYER_CONF = { initialPosition: {x: 1, y: 5}, sprite: '...', ...etc... };`
+1. Requirements re **OOP**:
```
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:320]:
```suggestion
1. Requirements regarding **OOP**:
```

File: docs/doc/code-review.md (hunk 9 above)
Comment (OleksiyRudenko) [reply docs/doc/code-review.md:321]: I'd cosider "doesn't make sense" rather than "against".
Q: Why `prototype.*` is still here? A: Legacy. The exisiting codebase is all `prototype.*`. Former Udacity theoretical part was about `prototype.*` only.

The progress in standards should be accounted for. 
I'd change this for the next course iteration. This will require to rework the original project (both codebase and documentation). Also `prototype.*` allows to tweak game engine re-defining its internals in a different file (e.g. `app.js`), which is not really possible with `class` syntax. Although no one used this opportunity yet anyway.

In a nutshell, I'd postpone this change at least until current Stage 0 end. It will also need some thoughtful effort at the expense of reviews or other improvements.

File: docs/doc/code-review.md (hunk 7 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:331]:
```suggestion
      (e.g. `update` method does only character coordinates update and
```

File: docs/doc/code-review.md [hunk 12]
```diff
@@ -248,80 +280,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
+      (e.g. there is `Character` base class, which is extended by `Enemy` and `Player` classes)
+   1. class extension is implemented using `Subclass.prototype = Object.create(Superclass.prototype)`,
+      not `Subclass.prototype = new Superclass(params);`;
+      [Useful resource](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)
+   1. classes do not refer to any global variables, like global variable `player`, which is an instance of `Player` class
+      (referring to global constants and globals provided by the gaming platform like `Resources` is OK);
+      Hint: pass instance of a game object (or objects) as an argument to other game objects they need to interact with
+   1. Separation of Concerns principle is followed
+      (e.g. `update` method does only rendering and
+      doesn't **contain** any inline code to check e.g. collisions)
```
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:332]:
```suggestion
      doesn't contain any **inline** code to check e.g. collisions; calling other methods from `update` is legitimate)
```

File: docs/doc/code-review.md (hunk 8 above)
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:344]:
```suggestion
   1. Implement base classes for child classes to inherit shared properties from.
```

**Comment by OleksiyRudenko** [comment]
@artem-trubin thank you for your review. Please review the code suggestions above.
I will merge those once we are aligned to reduce the review scope to the outstanding items.

There are two outstanding items:
- links to habr
- `prototype.*` vs `class`

Let's go on with these after smaller changes get committed.

File: docs/doc/code-review.md (hunk 2 above)
Comment (OleksiyRudenko) [reply docs/doc/code-review.md:37]: I like `descriptive` more as it more specific than `proper`.

**Review by OleksiyRudenko — COMMENTED** [review]
The suggestions below
* remove habr translation links
* revert `prototype` requirement on Frogger

**Review by suchov — APPROVED** [review]

File: docs/doc/code-review.md [hunk 13]
```diff
@@ -47,20 +53,39 @@ let myVar, myOtherVar;
```
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:53]:
```suggestion
_(although some conventional shorthands are OK; e.g. `i`, `j` to denote array indexes, `x`, `y` to denote coordinates and other similar cases)_
```

Comment (OleksiyRudenko) [inline docs/doc/code-review.md:53]:
```suggestion
_(although some conventional shorthands are OK; e.g. `i` to when looping through array by index, `x`/`y` to denote coordinates and other similar cases)_
```

**Review by OleksiyRudenko — COMMENTED** [review]
Minor update on variable naming to avoid naming like `axisX` or `valueY` in submitted code.

File: docs/doc/code-review.md [hunk 14]
```diff
@@ -25,11 +30,10 @@ PRs should have proper name (as per task name), should contain link to working d
...
-- Please advise your peer to fix irregular indentations and remove redundant empty lines.
-- Please advise your peer to put newline at the end of every file.
-- Unnecessary comments should be avoided. Advise your peer to find a way to express their intent through expressive variable names or by abstracting part of the code into properly named function.
-- Magic numbers are [bad](https://stackoverflow.com/questions/47882/what-is-a-magic-number-and-why-is-it-bad)
-- Add an empty lines at the end of each file. [Reason](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline?noredirect=1&lq=1)
+- Fix irregular indentations and remove redundant empty lines.
+- Put an empty line at the end of every file. [Reason](https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline?noredirect=1&lq=1). Tune your code editor's settings so it does this for you.
+- Unnecessary comments should be avoided. Find a way to express the intent through descriptive variable names or by abstracting part of the code into properly named function. Comments, if there are any, should explain "why", not "what".
```
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:35]:
```suggestion
- Unnecessary comments should be avoided. Find a way to express the intent through descriptive variable names or by abstracting parts of the code into properly named functions. Comments, if there are any, should explain "why", not "what".
```

**Review by OleksiyRudenko — COMMENTED** [review]
Missed a comment re function**s**

File: docs/doc/code-review.md [hunk 15]
```diff
@@ -248,80 +278,89 @@ Don't use inline style changing - `element.style`. In most cases this is a bad a
...
       than inheritance only
-
-Bonus:
-
-- [ ] `toString` magic method; when implemented `print(inhabitant)`
+1. Bonus:
+   1. `toString` magic method; when implemented `print(inhabitant)`
       does the job as `.toString` is called implicitly
-- [ ] `this.constructor.name`; when used no need to store `species` property
+   1. `this.constructor.name`; when used properly there is no need to specify `species` property explicitly
```
Comment (OleksiyRudenko) [inline docs/doc/code-review.md:359]:
```suggestion
   1. when `this.constructor.name` is used properly there is no need to specify `species` property explicitly
```

**Comment by OleksiyRudenko** [comment]
Added all suggestions.
Removed requirement on `prototype` syntax for Frogger.
